phi_four.py

```diff
@@ -181,82 +107,119 @@ def train(model, action, epochs):
+    action: Module
+        the action upon which the model was trained, used to calculate the
+        acceptance condition
+    target_length: int
+        the desired number of states to generate from the model
+    n_large: int (default 20000)
+        the number of total states to generate from the model per batch, set
+        to a reasonable default value. setting this number excessively high can
+        cause slow down due to running out of memory
+
+    Returns
+    -------
+    sample: torch.Tensor
+        a sample of states from model, size = (target_length, model.size_in)
+
     """
-    n_units = model.size_in
-    with torch.no_grad(): # don't want gradients being tracked in sampling stage
-        z = torch.randn((n_large, n_units)) # random z configurations
-        phi = model.inverse_map(z) # map using trained model to phi
-        log_ptilde = model.forward(phi) # log of probabilities of generated phis using trained model
-        S = action(phi)
-        chain_len = 0 # intialise current chain length
-        sample_distribution = torch.Tensor(target_length, n_units) # intialise tensor to store samples
-        accepted, rejected = 0,0 # track accept/reject statistics
-        i = RNG.randint(n_large) # random index to start sampling from configurations tensor
-        used = [] # track which configurations have been added to the chain
-        used.append(i) 
-        while chain_len < target_length:
-            j = RNG.randint(n_large) # random initial phi^j for update proposal
-            while j in used: # make sure we don't pick a phi we have already used
-                j = RNG.randint(n_large)
-            exponent = log_ptilde[i] - S[j] - log_ptilde[j] + S[i]
-            P_accept = exp(float(exponent)) # much faster if you tell it to use a float
-            A = min(1, P_accept) # faster than np.min and torch.min
-            u = RNG.uniform() # pick a random u for comparison
-            if u <= A:
-                sample_distribution[chain_len,:] = phi[i,:] # add to chain if accepted
-                chain_len += 1 
-                i = j # update i for next proposal
-                used.append(i)
+    accepted = 0
+    rejected = 0
+    batches = []
+
+    while accepted < target_length:
+        with torch.no_grad(): # don't track gradients
+            z = torch.randn((n_large, model.size_in)) # random z configurations
+            phi = model.inverse_map(z) # map using trained model to phi
+            log_ptilde = model(phi)
+        accept_reject = torch.zeros(n_large, dtype=bool)
+
+        log_ratio = log_ptilde + action(phi)
+        if ~np.isfinite(exp(float(min(log_ratio) - max(log_ratio)))):
+            raise ValueError("could run into nans")
+
+        if not batches: # first batch
+            log_ratio_i = log_ratio[0]
+            start = 1
+        else:
+            start = 0 # keep last log_ratio_i and start from 0
+
+        for j in range(start, n_large):
```

Comment:
I think there was no need to have the slightly complex random phi, instead can just work down the generate sample in order. This then allows us to easily 'batch' the sampling for larger samples. Now I can get approx 1e6 samples in 100 seconds